book/docs/lectures/Lecture4.md

```diff
@@ -0,0 +1,243 @@
+```haskell
+myContract2 :: Contract () BlockchainActions Void ()
+myContract2 = Contract.handleError
+  (\err -> Contract.logError $ "Caught error: " ++ unpack err)
+  myContract1
+```
+We have chosen the `Void` error type of this second contract. As this data type has no inhabitant in Haskell, this means that this contract can not have any errors. We do this in order to show that the error from the first contract is indeed handled. As we can see, the function that handles the error just take this error and we unpack it to convert it to the `String` type (at this moment it is of type `Text`, as we declared it with the parameter `e` on `myContract1`). Then we append it to the string message and log it as an error to the console with `Contract.logError`. Finally, if we simulate this contract using the trace as we have already learned, we are shown a message like the next one:
+```
+Prelude Week04.Contract> test2
+Slot 00000: TxnValidate af5e6d25b5ecb26185289a03d50786b7ac4425b21849143ed7e18bcd70dc4db8
+Slot 00000: SlotAdd Slot 1
+Slot 00001: 00000000-0000-4000-8000-000000000000 {Contract instance for wallet 1}:
+  Contract instance started
+Slot 00001: *** CONTRACT LOG: "Caught error: BOOM!"
+Slot 00001: 00000000-0000-4000-8000-000000000000 {Contract instance for wallet 1}:
+  Contract instance stopped (no errors)
+Slot 00001: SlotAdd Slot 2
+Final balances
+Wallet 1:
+    {, ""}: 100000000
+Wallet 2:
+    {, ""}: 100000000
+Wallet 3:
+    {, ""}: 100000000
+Wallet 4:
+    {, ""}: 100000000
+Wallet 5:
+    {, ""}: 100000000
+Wallet 6:
+    {, ""}: 100000000
+Wallet 7:
+    {, ""}: 100000000
+Wallet 8:
+    {, ""}: 100000000
+Wallet 9:
+    {, ""}: 100000000
+Wallet 10:
+    {, ""}: 100000000
+```
+where, as we see, the message is no longer an error but a log from the contract that informs us about the error. As the error data type of `myContract2` is of type `Void` we can be sure it was handled.
+
+## The Schema parameter: s
+We can define a custom set of contract actions by adding this actions to the `BlockchainActions` type. For example, let us say we want to add and endpoint called 'foo'. We just need to give a pseudonym to the set of action data types like this:
+```haskell
+{-# LANGUAGE DataKinds     #-}
+{-# LANGUAGE TypeOperators #-}
+-- ^ Add extensions on the top of the module file
+
+type MySchema = BlockchainActions .\/ Endpoint "foo" Int
+```
+In the last line we define the type of the set of actions and we call it `MySchema`. Then we use the operator `.\/`, which acts on types, not on values, to "add" the endpoints that we want, in this case the `foo` endpoint. The first argument to `Endpoint` is a type level string which represents the name of the endpoint, and the second argument is the parameter type (which type of value this endpoint takes).
```

Comment:
```suggestion
In the last line we've defined the type of the set of actions and we've called it `MySchema`. Then we used the operator `.\/`, which acts on types, not on values, to "add" the endpoints that we want, in this case the `foo` endpoint. The first argument to `Endpoint` is a type level string which represents the name of the endpoint, and the second argument is the parameter type (which type of value this endpoint takes).
```

Reply:
```suggestion
In the last line we have defined the type of the set of actions and we have called it `MySchema`. Then we used the operator `.\/`, which acts on types, not on values, to "add" the endpoints that we want, in this case the `foo` endpoint. The first argument to `Endpoint` is a type level string which represents the name of the endpoint, and the second argument is the parameter type (which type of value this endpoint takes).
```

I am trying not to use contractions, but I agree on the verb tense